blockntfns/manager.go

```diff
@@ -0,0 +1,346 @@
+// be delivered through the Notifications channel within the Subscription
+// returned. A Cancel closure is also provided, in the event that the client
+// wishes to no longer receive any notifications.
+func (m *SubscriptionManager) NewSubscription(bestHeight uint32) (*Subscription,
+	error) {
+
+	// We'll start by constructing the internal messages that the
+	// subscription handler will use to register the new client.
+	sub := &newSubscription{
+		id:         atomic.AddUint64(&m.subscriberCounter, 1),
+		ntfnChan:   make(chan BlockNtfn, 20),
+		ntfnQueue:  NewConcurrentQueue(20),
+		bestHeight: bestHeight,
+		errChan:    make(chan error, 1),
+		quit:       make(chan struct{}),
+	}
+
+	// We'll start the notification queue now so that it is ready in the
+	// event that a backlog of notifications is to be delivered.
+	sub.ntfnQueue.Start()
+
+	// We'll also start a goroutine that will attempt to consume
+	// notifications from this queue by delivering them to the client
+	// itself.
+	sub.wg.Add(1)
+	go func() {
+		defer sub.wg.Done()
+
+		for {
+			select {
+			case ntfn, ok := <-sub.ntfnQueue.ChanOut():
+				if !ok {
+					return
+				}
+
+				select {
+				case sub.ntfnChan <- ntfn.(BlockNtfn):
+				case <-sub.quit:
+					return
+				case <-m.quit:
+					return
+				}
+			case <-sub.quit:
+				return
+			case <-m.quit:
+				return
+			}
+		}
+	}()
+
+	// Now, we can deliver the notification to the subscription handler.
+	select {
+	case m.newSubscriptions <- sub:
+	case <-m.quit:
+		sub.ntfnQueue.Stop()
+		return nil, ErrSubscriptionManagerStopped
+	}
+
+	// It's possible that the registration failed if we were unable to
+	// deliver the backlog of notifications, so we'll make sure to handle
+	// the error.
+	select {
+	case err := <-sub.errChan:
+		if err != nil {
+			sub.ntfnQueue.Stop()
+			return nil, err
+		}
+	case <-m.quit:
+		sub.ntfnQueue.Stop()
+		return nil, ErrSubscriptionManagerStopped
+	}
+
+	// Finally, we can return to the client with its new subscription
+	// successfully registered.
+	return &Subscription{
+		Notifications: sub.ntfnChan,
+		Cancel: func() {
+			m.cancelSubscription(sub)
+		},
+	}, nil
+}
+
+// handleNewSubscription handles a request to create a new block subscription.
+func (m *SubscriptionManager) handleNewSubscription(sub *newSubscription) error {
+	log.Infof("Registering block subscription: id=%d", sub.id)
+
+	// We'll start by retrieving a backlog of notifications from the
+	// client's best height.
+	blocks, currentHeight, err := m.ntfnSource.NotificationsSinceHeight(
+		sub.bestHeight,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve blocks since height=%d: "+
+			"%v", sub.bestHeight, err)
+	}
+
+	// We'll then attempt to deliver these notifications.
+	log.Debugf("Delivering backlog of block notifications: id=%d, "+
+		"start_height=%d, end_height=%d", sub.id, sub.bestHeight,
+		currentHeight)
+
+	for _, block := range blocks {
+		m.notifySubscriber(sub, block)
```

Comment:
You could use a timeout case in the `notifySubscriber()` method to return an error; otherwise, it'll just sit there and wait for a deadlocked subscriber.

Reply:
This won't actually block since it's being sent into the queue and not the buffered channel.

> should make this method return an error, so we can know whether the notification was delivered.

Why would it matter if the notification was actually delivered from the PoV of the manager? In this case it won't be delivered either because the client canceled its subscription or the manager has been shut down.

Reply:
That's already the case? The `select` statement handles the manager's and subscriber's `quit chan`s.